core/src/main/java/org/bitcoinj/core/Peer.java

```diff
@@ -169,7 +170,10 @@ public GetDataRequest(Sha256Hash hash) {
 
     // Outstanding pings against this peer and how long the last one took to complete.
     private final ReentrantLock pingIntervalsLock = new ReentrantLock();
-    @GuardedBy("pingIntervalsLock") private Duration[] pingIntervals = null;
+    @GuardedBy("pingIntervalsLock") private final Deque<Duration> pingIntervals = new ArrayDeque<>(PING_MOVING_AVERAGE_WINDOW);
+    @GuardedBy("pingIntervalsLock") private volatile Duration lastPing = null;
+    @GuardedBy("pingIntervalsLock") private volatile Duration avgPing = null;
```

Comment:
Is this `@GuardedBy` actually right? I mean, like you said in the getters it's now unguarded. And it's `volatile`.

Reply:
Well, it's guarded for writes.

Reply:
But then the read access would be violating of the locking policy declared by `@GuardedBy`, doesn't it?

I think the idea of the annotation is that it can be checked by build tools/IDEs.

Reply:
I wrote the above under the assumption that the two fields are independent of each other.

If it needs to be ensured that `lastPing` and `avgPing` are consistent with each other, we'd need to switch to a different model anyway. But I'd say that would be overkill.